content/docs/2.7/scalers/azure-pipelines.md

```diff
@@ -37,6 +39,7 @@ triggers:
 - `personalAccessTokenFromEnv` - Name of the environment variable that provides the personal access token (PAT) for Azure DevOps. Learn more about how to create one [in the official docs](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=preview-page).
 - `targetPipelinesQueueLength` - Target value for the amount of pending jobs in the queue to scale on. (Default: `1`, Optional)
   - Example - If one pod can handle 10 jobs, set the queue length target to 10. If the actual number of jobs in the queue is 30, the scaler scales to 3 pods.
+- `parent` - Put the name of the ADO agent that matched the ScaledJob. e.g. mavenagent-scaledjob may have an initial deployment called "mavenagent-keda-template"; this is the deployment that is made offline.
```

Comment:
Frankly, I don't really understand what is expected here or how it works.

Reply:
Isn't it as simple to just provide a list of capabilities?

Reply:
I think that capabilities is the best approach because in fact, it's MS approach

Reply:
My first couple of attempts both use capabilities in the Job. This method means there is then an extra API call to check if the agent can fulfil the demand. Inside the actual JobRequest along with demands is a field called matchedAgent which are the agents that ADO has automatically matched. This means ADO does the evaluation for us and already tells us what can fulfil it.
The other part is this separates the "ADO" specific stuff into the Dockerfile and has KEDA just doing the KEDA stuff.

Reply:
This is the other way of doing it and considering ADO is rate limited could cause more issues IMHO
`				var demands = ""
				for _, v := range v["demands"].([]interface{}) {
					demands = demands + "," + strings.Trim(v.(string), " ")
				}
				// tidy for querysting and make url safe
				demands = strings.ReplaceAll(strings.TrimLeft(demands, ","), " ", "%20")

				// determine if this agent can satisfy the job
				urlAgent := fmt.Sprintf("%s/_apis/distributedtask/pools/%d/agents?agentName=%s&demands=%s", s.metadata.organizationURL, s.metadata.poolID, s.metadata.parent, demands)
				bodyAgent, err := getAzurePipelineRequest(ctx, urlAgent, s.metadata, s.httpClient)
				if err != nil {
					return -1, err
				}

				var resultAgent map[string]interface{}
				err = json.Unmarshal(bodyAgent, &resultAgent)
				if err != nil {
					return -1, err
				}

				countAgent := int(resultAgent["count"].(float64))
				if countAgent > 0 {
					count++
				}`

Reply:
I'd add a list of capabilities inside metadata and I'd check them parsing the response. I have done the request and this is a sample of the response:
```json
{
   "count":1,
   "value":[
      {
         "requestId":1493,
         "queueTime":"............................",
         "serviceOwner":"...........................",
         "hostId":"..............................",
         "scopeId":".......................",
         "planType":"Build",
         "planId":"..................",
         "jobId":"..................",
         "demands":[
            "Agent.Version -gtVersion 2.144.0"
         ],
         "matchedAgents":[
            {
               "_links":{
                  "self":{
                     "href":"https://dev.azure.com/...................../_apis/distributedtask/pools/13/agents/161"
                  },
                  "web":{
                     "href":"https://dev.azure.com/...................../_settings/agentpools?view=jobs&poolId=13&agentId=161"
                  }
               },
               "id":161,
               "name":"test-deployment-8474f4bd57-2gcj2",
               "version":"2.196.2",
               "enabled":true,
               "status":"offline",
               "provisioningState":"Provisioned"
            }
         ],
         "definition":{
            "_links":{
               "web":{
                  "href":"https://dev.azure.com/..................................../_build/definition?definitionId=29"
               },
               "self":{
                  "href":"https://dev.azure.com/...................................../_apis/build/Definitions/29"
               }
            },
            "id":29,
            "name":"KEDA-CI"
         },
         "owner":{
            "_links":{
               "web":{
                  "href":"https://dev.azure.com/................................/_build/results?buildId=433"
               },
               "self":{
                  "href":"https://dev.azure.com/............................./_apis/build/Builds/433"
               }
            },
            "id":433,
            "name":"433"
         },
         "data":{
            "ParallelismTag":"Public",
            "IsScheduledKey":"False"
         },
         "poolId":13,
         "orchestrationId":"eee66f1e-5e66-4003-a1c3-a9da4b6e844f.job_1.__default",
         "matchesAllAgentsInPool":true,
         "priority":0
      }
   ]
}
```

I'm not 100% sure because I couldn't test it but I think that each element (job) contains the demands under the section `demands`. Maybe we could iterate the result and calculate how many pending jobs demand the supported capabilities.
This approach doesn't add any extra request to the Api so the limit rate should not be a problem.
WDYT?

Reply:
That was actually where I started too; and it does work. The problem is that you have to check the Job's demands from the request against the named demands in your metadata and you cannot be sure they are accurate. If you log into `ADO -> Project Settings -> Agent Pools -> PoolName -> Agents -> AgentName -> Capabilities`.
This is the list of demands that the Agent actually supports. These cannot all be codified into your ScaledJob metadata as anytime an agent's environment variables change, so do the demands.
This does work if you have a specific named demand to target an agent, such as the ScaledJob has a demand called `"keda-maven"` and your job has a specific demand of the same. Evaluating all the others is the issue.
**This was attempt 1.**

**Attempt 2:**
As above, but using the Job Request to ask ADO which agents support the requested demands; this used the parent methodology as in the PR but adds an extra API call to interrogate ADO.

**Attempt 3:**
ADO has already told you the list of agents that can do the job. This is in the matchedAgents field. As long as 1 of those agents can be related to the ScaledJob (I use a template agent and the parent metadata) then you know which to spin up.

Reply:
I get your point, but I still have doubts about this approach.
Any thoughts @kedacore/keda-contributors ?

Reply:
What about using deployments instead of statefulset for agents? I mean, the agent register itself using the hostname if I'm not wrong. How will this affect in that case?

Reply:
any  update here?

Reply:
good point, the environment variable "AZP_NAME" has to be set. I'll update the docs

Reply:
Any thoughts https://github.com/orgs/kedacore/teams/keda-contributors ?

Reply:
Any thoughts @kedacore/keda-contributors ?

Reply:
I think that's a fair requirement to have `AZP_NAME`. However, does that have to be on the agent running (end-user deployment) or something we can do from KEDA?

I think it's the first but double-checking.

Reply:
Its the agent as that is what get passed into the job call that ADO actually makes and it reads it from its registered list inside the ADO project